pkg/queue/breaker.go

```diff
@@ -60,12 +65,103 @@ func (b *Breaker) Maybe(thunk func()) bool {
 	case b.pendingRequests <- t:
 		// Pending request has capacity.
 		// Wait for capacity in the active queue.
-		b.activeRequests <- t
+		b.sem.Get()
 		// Defer releasing capacity in the active and pending request queue.
-		defer func() { <-b.activeRequests; <-b.pendingRequests }()
+		defer func() { b.sem.Put(); <-b.pendingRequests }()
 		// Do the thing.
 		thunk()
 		// Report success
 		return true
 	}
 }
+
+// NewSemaphore creates a semaphore with the desired maximal capacity and initial capacity
+func NewSemaphore(maxCapacity, initialCapacity int32) *Semaphore {
+	if initialCapacity < 0 || initialCapacity > maxCapacity {
+		panic(fmt.Sprintf("Initial capacity must be between 0 and maximal capacity. Got %v.", initialCapacity))
+	}
+	waiters := make(chan token, maxCapacity)
+	reducers := make(chan token, maxCapacity)
+	mainQueue := make(chan token, maxCapacity)
+	sem := Semaphore{waitersQueue: waiters, reducersQueue: reducers, mainQueue: mainQueue}
+	if initialCapacity > 0 {
+		//for i := int32(0); i < initialCapacity; i++ {
+		//	sem.AddCapacity()
+		//}
+		sem.AddCapacity(initialCapacity)
+	}
+	return &sem
+}
+
+// Semaphore is an implementation of a semaphore based on Go channels
+type Semaphore struct {
+	mux           sync.Mutex
+	waitersQueue  chan token
+	reducersQueue chan token
+	mainQueue     chan token
+	available     int32
+	capacity      int32
+	waiters       int32
+	reducers      int32
+	token         token
+}
```

Comment:
To quickly summarize the need for a slightly more complex structure here (and to spare other reviewers that work):

The general problem of a resizable Semaphore is taking capacity away. That is: If there are a lot of requests in flight, we might not be able to take as many tokens out of the queue as we want. There we face two alternatives:

1. Make ReduceCapacity blocking until enough tokens available.
2. Defer the capacity reductions across all `Put` calls by deciding if we want to put the token back on the queue.

We're trying to go for option 2 here but I think this can be simplified a bit, only needing one channel and one number. 3 Channels are quite daunting. I've engaged a 1:1 discussion with @vvraskin (as that is much more productive for a problem like this). Will report back on the outcome ASAP.

Option 1 *could* be a little more dangerous as we could potentially run a lot of goroutines in parallel while Option 2 has the tradeoff of needing locking or atomic magic.